### PR TITLE
Cassandra: check `isFullyFetched` before completion

### PR DESCRIPTION
**Pull request by anzecesar**
## Purpose

Under some conditions, the source completed before all results were loaded from the result set.

Now it only completes when rs is fully fetched and all elements were emitted. Before it was completing in case where the current fetched buffer was empty.

## References

References https://github.com/akka/alpakka/issues/1934

## Changes

* only completes the stream when `rs.isFullyFetched`
* if just the currently loaded set is empty, it fetches more (an edge case)

## Background Context

I observed this issue where fetchSize was set relatively low and the
downstream consumer was fast.

### CODE REVIEW

**Comment by anzecesar** [comment]
@ennru I can try, but this was already hard to reproduce. Are there any existing tests with cassandra already?

**Comment by ennru** [comment]
Yes, they use Cassandra via Docker compose.

**Comment by anzecesar** [comment]
@ennru Where can I find them? There's nothing in https://github.com/akka/alpakka/tree/master/cassandra/src/test

**Comment by ennru** [comment]
https://github.com/akka/alpakka/blob/master/cassandra/src/test/scala/docs/scaladsl/CassandraSourceSpec.scala

**Comment by anzecesar** [comment]
@ennru Tnx. For some reason I assumed the purpose of that was more or less just for the docs and haven't even considered it a working spec 🤷‍♂️. Well, we all know where assumption leads 😂.

I'll try to add a test case for this although I suspect it will be at best very flaky.

**Comment by anzecesar** [comment]
I finally got some time to work on this today and I was not able to reproduce this locally with a spec. The conditions under which I originally encountered the issue were reading ~10k rows containing quite large jsons (around 150 mb for the whole dataset). I'm not exactly sure what exactly led to the buffer being exhausted prematurely.

There is however an exiting test, that generally ensures this doesn't break any functionality: https://github.com/akka/alpakka/blob/aafad95ffa49a08c99342bd64752a2e4f1cae399/cassandra/src/test/scala/docs/scaladsl/CassandraSourceSpec.scala#L107

File: cassandra/src/main/scala/akka/stream/alpakka/cassandra/impl/CassandraSourceStage.scala [hunk 1]
```diff
@@ -36,18 +36,22 @@ import scala.util.{Failure, Success, Try}
       }(exec)
     }
 
+    private[this] def fetch(rs: ResultSet): Unit = {
+      isFetching = true
+      // fetch next page
+      val gFut = rs.fetchMoreResults()
+      val exec = materializer.executionContext
+      GuavaFutures.invokeTryCallback(gFut, exec)(futFetchedCallback)
```
Comment (chbatey) [inline cassandra/src/main/scala/akka/stream/alpakka/cassandra/impl/CassandraSourceStage.scala]: should this be using an async call back?

Comment (ennru) [reply cassandra/src/main/scala/akka/stream/alpakka/cassandra/impl/CassandraSourceStage.scala]: It looks OK to me, it is only called within synchronous code. 
`tryPushAfterFetch` is called via the async handler `futFetchedCallback` when the future returns.

Comment (anzecesar) [reply cassandra/src/main/scala/akka/stream/alpakka/cassandra/impl/CassandraSourceStage.scala]: Yes, I just refactored it out into a method to not duplicate code.

**Comment by ennru** [comment]
Thank you for this fix.

**Pull request merged**